p2p/peer_behaviour.go

```diff
@@ -0,0 +1,89 @@
+package p2p
+
+type ErrorBehaviourPeer int
+
+const (
+	ErrorBehaviourUnknown = iota
+	ErrorBehaviourBadMessage
+	ErrorBehaviourMessageOutofOrder
+)
+
+type GoodBehaviourPeer int
+
+const (
+	GoodBehaviourVote = iota
+	GoodBehaviourBlockPart
+)
+
+// PeerBehaviour provides an interface for reactors to signal the behaviour
+// of peers synchronously to other components.
+type PeerBehaviour interface {
+	Behaved(peer Peer, reason GoodBehaviourPeer)
+	Errored(peer Peer, reason ErrorBehaviourPeer)
+}
+
+type switchedPeerBehaviour struct {
+	sw *Switch
+}
+
+func (spb *switchedPeerBehaviour) Errored(peer Peer, reason ErrorBehaviourPeer) {
+	spb.sw.StopPeerForError(peer, reason)
+}
+
+func (spb *switchedPeerBehaviour) Behaved(peer Peer, reason GoodBehaviourPeer) {
+	spb.sw.MarkPeerAsGood(peer)
+}
+
+func NewswitchedPeerBehaviour(sw *Switch) PeerBehaviour {
+	return &switchedPeerBehaviour{
+		sw: sw,
+	}
+}
+
+type ErrorBehaviours map[Peer][]ErrorBehaviourPeer
+type GoodBehaviours map[Peer][]GoodBehaviourPeer
+
+type IStorePeerBehaviour interface {
```

Comment:
I assume that you want to expose this for consistent testing tooling in packages that depend on `PeerBehaviour`. Is that correct? If so I welcome this effort as we've seen with `Peer` we otherwise might end up with heaps of local `mock` implementations, which leads to unnecessary duplicated effort esp. with more complext types and harder to navigate test code for the uninitiated (often even the initiated).

Why is the name prefixed with an `I`?

Also embedding the `PeerBehaviour` interface in here makes it less flexible. I'd omit it and either have the caller construct a combined interface a la:

``` go
type myLocalTestBehaviour interface {
    p2p.PeerBehaviour
    p2p.StorePeerBehaviour
}
```
or expose a combined one from here (not as elegant).

At first I was on the fence about the name and was tempted to say we should signal more strongly that it is intended to be used in testing, but that might not be true.  :>